### PR TITLE
Lower SSE compare scalar and test nodes

### LINKED ISSUE #17073: CQ issues with HW intrinsics that return bool
HW intrinsics that produce a bool value need special handling when the value they produce is used by a conditional branch. As is now the intrinsic node produces the value in a register and then an additional test is required to check for false/true and branch.
For example
```C#
return Sse2.CompareEqualOrderedScalar(x, y) ? 4 : 2;
```
generates
```asm
       C4E1792FC1           vcomisd  xmm0, xmm1
       0F9BC2               setpo    dl
       0F94C0               sete     al
       22C2                 and      al, dl
       0FB6D0               movzx    rdx, al
       85D2                 test     edx, edx
       7506                 jne      SHORT G_M55886_IG04
       B802000000           mov      eax, 2
G_M55886_IG03:
       C3                   ret
G_M55886_IG04:
       B804000000           mov      eax, 4
G_M55886_IG05:
       C3                   ret
```
but it should generate something like
```asm
G_M55886_IG02:
       C4E1792EC1           vcomisd xmm0, xmm1
       7A02                 jpe      SHORT G_M55886_IG03
       7406                 je       SHORT G_M55886_IG05
G_M55886_IG03:
       B802000000           mov      eax, 2
G_M55886_IG04:
       C3                   ret
G_M55886_IG05:
       B804000000           mov      eax, 4
G_M55886_IG06:
       C3                   ret
```
As far as I can tell this affects pretty much all HW intrinsics that return bool. Except `IsSupported`, of course.

category:cq
theme:intrinsics
skill-level:expert
cost:medium

### LINKED ISSUE #21247: !Avx.Test{Z,C} suboptimal codegen
`!Avx.TestZ` and `!Avx.TestC` produce suboptimal code, can be shown with a simple repro:

```c#
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

namespace ConsoleApp3
{
    class Program
    {
        static int Main(string[] args)
        {
            Vector256<sbyte> a = Avx.SetAllVector256<sbyte>(-42);
            Vector256<sbyte> b = Avx.SetAllVector256<sbyte>(42);

            bool res = Foo(a, b);

            return res ? 1 : 0;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        private static bool Foo(Vector256<sbyte> a, Vector256<sbyte> b)
        {
            return !Avx.TestZ(a, b);
        }
    }
}
```
results in (method `Foo`):
```asm
G_M7898_IG01:
       C5F877               vzeroupper 
       6690                 nop      

G_M7898_IG02:
       C4E17D10442408       vmovupd  ymm0, ymmword ptr[rsp+08H]
       C4E27D17442428       vptest   ymm0, ymmword ptr[rsp+28H]
       0F94C0               sete     al
       0FB6C0               movzx    rax, al
       85C0                 test     eax, eax
       0F94C0               sete     al
       0FB6C0               movzx    rax, al

G_M7898_IG03:
       C5F877               vzeroupper 
       C3                   ret      
```
Even withoud `NoInlining` this not ideal pattern can be seen.

Instead of the additional `test`-part, `setne` would be enough:
```diff
G_M7898_IG01:
       C5F877               vzeroupper
       6690                 nop

G_M7898_IG02:
       C4E17D10442408       vmovupd  ymm0, ymmword ptr[rsp+08H]
       C4E27D17442428       vptest   ymm0, ymmword ptr[rsp+28H]
-      0F94C0               sete     al
-      0FB6C0               movzx    rax, al
-      85C0                 test     eax, eax
-      0F94C0               sete     al
+      0F94C0               setne    al
       0FB6C0               movzx    rax, al

G_M7898_IG03:
       C5F877               vzeroupper
       C3                   ret
```
`COMPlus_TieredCompilation=0` is set, so it is tier-1 code.

Codegen for the "positive" case (i.e. `Avx.TestZ` and `Avx.TestC`) is optimal.

category:cq
theme:intrinsics
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by mikedn**
This prevents the materialization of a bool value when the intrinsic node is used by a JTRUE node:
```diff
        C5F91001             vmovupd  xmm0, xmmword ptr [rcx]
        C4E2791702           vptest   xmm0, xmmword ptr [rdx]
-       0F97C0               seta     al
-       0FB6C0               movzx    rax, al
-       85C0                 test     eax, eax
-       750C                 jne      SHORT G_M18062_IG04
-       E8E1E2FFFF           call     Program:False():bool
+       770C                 ja       SHORT G_M18062_IG04
+       E8E9E5FFFF           call     Program:False():bool
        90                   nop      
```

Diff from the autogenerated test: https://gist.github.com/mikedn/35f8e21be5c9da22ade98241598b5243

Fixes #17073
Fixes #21247

### CODE REVIEW

File: src/jit/hwintrinsiclistxarch.h [hunk 1]
```diff
@@ -646,6 +646,15 @@ HARDWARE_INTRINSIC(POPCNT_PopCount,                                 "PopCount",
...
+HARDWARE_INTRINSIC(SSE41_PTEST,                                     "PTEST",                                        SSE41,        -1,              16,           2,     {INS_ptest,             INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_invalid,        INS_invalid},           HW_Category_SimpleSIMD,             HW_Flag_NoRMWSemantics)
+HARDWARE_INTRINSIC(AVX_PTEST,                                       "PTEST",                                        AVX,          -1,               0,           2,     {INS_ptest,             INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_ptest,          INS_vtestps,        INS_vtestpd},           HW_Category_SimpleSIMD,             HW_Flag_NoRMWSemantics)
```
Comment (fiigii) [inline src/jit/hwintrinsiclistxarch.h]: Perhaps we can give a new category (or flag) to these internal intrinsic, so that the codegen can explicitly check the category rather than rely on type == VIOD.

Comment (mikedn) [reply src/jit/hwintrinsiclistxarch.h]: A new category may be useful for other reasons but checking for TYP_VOID seems normal to me - nodes that do not generate a value need special handling.

File: src/jit/gentree.h [hunk 2]
```diff
@@ -5626,8 +5626,8 @@ struct GenCondition
         C    = Unsigned | S,    // = 14
         NC   = Unsigned | NS,   // = 15
                                 
-        FEQ  = Float | EQ,      // = 16
-        FNE  = Float | NE,      // = 17
+        FEQ  = Float | 0,       // = 16
```
Comment (mikedn) [inline src/jit/gentree.h]: This was a mistake that slipped in with my JCC PR a while ago. Since nothing used these constants before this went unnoticed.

File: src/jit/lowerxarch.cpp [hunk 3]
```diff
@@ -880,6 +902,114 @@ void Lowering::LowerSIMD(GenTreeSIMD* simdNode)
...
+            LowerHWIntrinsicCC(node, NI_SSE_COMISS, GenCondition::FLE);
+            break;
+        case NI_SSE_CompareGreaterThanOrderedScalar:
+            LowerHWIntrinsicCC(node, NI_SSE_COMISS, GenCondition::FGT);
+            break;
+        case NI_SSE_CompareGreaterThanOrEqualOrderedScalar:
+            LowerHWIntrinsicCC(node, NI_SSE_COMISS, GenCondition::FGE);
+            break;
+
+        case NI_SSE_CompareEqualUnorderedScalar:
```
Comment (mikedn) [inline src/jit/lowerxarch.cpp]: "Design by committee" intrinsic names. They give the impression that the condition is "unordered" but in fact it's the compare instruction itself that it "unordered" (aka quiet, aka non-signaling).

File: src/jit/lower.cpp [hunk 4]
```diff
@@ -2940,6 +2940,107 @@ GenTree* Lowering::LowerJTrue(GenTreeOp* jtrue)
...
+//
+// Notes:
+//     This simply replaces `node`'s use with an appropiate SETCC/JCC node,
+//     `node` is not actually changed, except by having its GTF_SET_FLAGS set.
+//     It's the caller's responsibility to change `node` such that it only
+//     sets the condition flags, without producing a boolean value.
+//
+GenTreeCC* Lowering::LowerNodeCC(GenTree* node, GenCondition condition)
+{
+    // Skip over a chain of EQ/NE(x, 0) relops. This may be present either
```
Comment (mikedn) [inline src/jit/lower.cpp]: The code in `LowerSIMD` that this replaces also checked for `EQ/NE(x, 1)`. I added that code in #14027 but now I can't find any use for that extra complexity. Even if both the JIT and the C# compiler sometimes assume 0/1 bools when performing some optimizations, I've never seen a case that involves _testing_ bools and assuming 0/1, testing is always 0/non-zero.

File: src/jit/lowerxarch.cpp [hunk 5]
```diff
@@ -801,77 +801,99 @@ void Lowering::LowerSIMD(GenTreeSIMD* simdNode)
...
-            //
-            // Try to transform JTRUE(EQ|NE(SIMD<OpEquality|OpInEquality>(x, y), 0|1)) into
-            // JCC(SIMD<OpEquality|OpInEquality>(x, y)). SIMD<OpEquality|OpInEquality>(x, y)
-            // is expected to set the Zero flag appropriately.
-            // All the involved nodes must form a continuous range, there's no other way to
-            // guarantee that condition flags aren't changed between the SIMD node and the JCC
-            // node.
-            //
+        simdNode->gtType = TYP_VOID;
+        simdNode->ClearUnusedValue();
```
Comment (mikedn) [inline src/jit/lowerxarch.cpp]: The old code didn't do this. Assuming one can produce such an unused SIMD node, that could have lead to an assert.

File: src/jit/lowerxarch.cpp (hunk 3 above)
Comment (tannergooding) [reply src/jit/lowerxarch.cpp]: Are you suggesting these be named `UnorderedCompareEqualScalar` instead? Or possibly some other name?

File: src/jit/lowerxarch.cpp [hunk 6]
```diff
@@ -880,6 +902,114 @@ void Lowering::LowerSIMD(GenTreeSIMD* simdNode)
...
+
+        case NI_SSE41_TestAllOnes:
+            LowerHWIntrinsicCC(node, NI_SSE41_PTEST, GenCondition::C);
+            break;
+
+        case NI_SSE41_TestC:
+            LowerHWIntrinsicCC(node, NI_SSE41_PTEST, GenCondition::C);
+            break;
+        case NI_SSE41_TestAllZeros:
+        case NI_SSE41_TestZ:
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/jit/lowerxarch.cpp]: These shouldn't be duplicated.

The former is `int _mm_test_all_zeros (__m128i a, __m128i mask)` and does:
```
IF (a[127:0] AND mask[127:0] == 0)
	ZF := 1
ELSE
	ZF := 0
FI
RETURN ZF
```

While the latter is `int _mm_testz_si128 (__m128i a, __m128i b)` and does:
```
IF (a[127:0] AND b[127:0] == 0)
	ZF := 1
ELSE
	ZF := 0
FI
IF ((NOT a[127:0]) AND b[127:0] == 0)
	CF := 1
ELSE
	CF := 0
FI
RETURN ZF
```

File: src/jit/lower.cpp (hunk 4 above)
Comment (tannergooding) [reply src/jit/lower.cpp]: > I've never seen a case that involves testing bools and assuming 0/1, testing is always 0/non-zero.

Not sure what you would consider "testing", but for C# code there are a number of places where the generated IL doesn't strictly do `0` or `not-zero` checks. For example, C# returns `false` for `bool == bool` where the underlying bits are not identical and a switch statement can hit the default case for a bool which is not 0/1. I believe the csharplang repo had a few more oddities commented at one time.

File: src/jit/lowerxarch.cpp (hunk 3 above)
Comment (mikedn) [reply src/jit/lowerxarch.cpp]: Yep, `UnorderedCompareEqualScalar` is a reasonable option because it's closer to the native intrinsic name and instruction name. Another option would be to not include the ordered ones because there's no obvious case for them in .NET.

File: src/jit/lowerxarch.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/lowerxarch.cpp]: I'm not sure what the description of the PTEST instruction that you show has to do with the intrinsic itself. The intrinsic does not return flags but a bool value and these 2 intrinsics are supposed to return `true` if the result is 0. That is, they return ZF. That's what the existing codegen does and that's what all native compiler do as well.

In Visual C++ `_mm_test_all_zeros` (and the other 2 SSE41 PTEST helpers) are just macros:
```C++
#define _mm_test_all_zeros(mask, val)      _mm_testz_si128((mask), (val))
#define _mm_test_all_ones(val) \
              _mm_testc_si128((val), _mm_cmpeq_epi32((val),(val)))
#define _mm_test_mix_ones_zeros(mask, val) _mm_testnzc_si128((mask), (val))
```
So yeah, they're duplicates.

File: src/jit/lower.cpp (hunk 4 above)
Comment (mikedn) [reply src/jit/lower.cpp]: By testing I mean something like
```C#
if (ptest(x, y)) ...
if (!ptest(x, y)) ...
bool b = ptest(x, y)...
bool b = !ptest(x, y)...
```
I've never seen the C# compiler generating anything other than 0 compares in all these cases. In particular, logical negation (the subject of #21247) is implemented using `ceq(x, 0)` or `brfalse`.

File: src/jit/lowerxarch.cpp (hunk 3 above)
Comment (tannergooding) [reply src/jit/lowerxarch.cpp]: > because there's no obvious case for them in .NET.

Right, and I believe last time I discussed this with @CarolEidt, we determined that there wasn't any pressing need to support floating-point exceptions (even as opt-in) for .NET. For `COMISS` and `UCOMISS` basically the only difference is that the former also throws for `QNaN` (when floating-point exceptions are enabled).

>  Another option would be to not include the ordered ones

I think a downside here is that we can't just call it `CompareEqualScalar` since you can't overload by return type (and `CMPSS` currently holds that name). Prefixing it with `Unordered` (and possibly `Ordered` for the COMISS` ones) also makes them less discoverable. Maybe `CompareOrderedEqualScalar` and `CompareUnorderedEqualScalar` would help keep them discoverable....

I'll split this out into a separate issue for further discussion.

File: src/jit/lowerxarch.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/jit/lowerxarch.cpp]: > In Visual C++ _mm_test_all_zeros (and the other 2 SSE41 PTEST helpers) are just macros

Ah, I see. I was going based-off the `Intel Intrinsics Guide` definitions which, at first glance, looks like they have differing semantics. However, I see that the code-segment dealing with the `CF` flag is superfluous.

We already had an issue around the naming of the `TestC`, `TestZ`, and `TestNotZAndNotC` methods, and I've updated it to suggest we just delete them given that they are duplicates: https://github.com/dotnet/corefx/issues/34882

[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/lowerxarch.cpp]: Eh, the intrinsic guide isn't exactly the brightest piece of documentation. The practically copied the description of `PTEST` and appended an easy to miss sentence at the end: "Return the ZF value.".

File: src/jit/lower.cpp (hunk 4 above)
Comment (tannergooding) [reply src/jit/lower.cpp]: > I've never seen the C# compiler generating anything other than 0 compares in all these cases

Ah, I think the one case where it does an explicit `ceq(x, 1)` (rather than `ceq(x, 0`, `brfalse`, or `brtrue`) is for switch statements. I doubt such a code-pattern is that common, however.

In this case:
```csharp
switch (condition)
{
    case false:
        Console.WriteLine("0");
        break;

    case true:
        Console.WriteLine("1");
        break;

    default:
        Console.WriteLine("?");
        break;
}
```

becomes:
```ilasm
        IL_0000: ldarg.1
        IL_0001: brfalse.s IL_0009

        IL_0003: ldarg.1
        IL_0004: ldc.i4.1
        IL_0005: beq.s IL_0014

        IL_0007: br.s IL_001f

        IL_0009: ldstr "0"
        IL_000e: call void [mscorlib]System.Console::WriteLine(string)
        IL_0013: ret

        IL_0014: ldstr "1"
        IL_0019: call void [mscorlib]System.Console::WriteLine(string)
        IL_001e: ret

        IL_001f: ldstr "?"
        IL_0024: call void [mscorlib]System.Console::WriteLine(string)
        IL_0029: ret
```

**Comment by tannergooding** [comment]
Is this currently pending anything or just review/sign-off? It would be nice to have for 3.0

**Comment by fiigii** [comment]
> It would be nice to have for 3.0

Second. I really want this opt in 3.0. 
@CarolEidt Could you please take a look at this PR?

**Comment by mikedn** [comment]
Hrm, this already picked up conflicts. I'm not sure what happened with the compare/test intrinsics, there were some API issues related to those.

**Comment by BruceForstall** [comment]
@mikedn Looks like this is ready to go now?

@tannergooding @CarolEidt @sandreenko How about another code review?

**Comment by mikedn** [comment]
@BruceForstall Yes, this is done.

File: src/jit/lowerxarch.cpp [hunk 7]
```diff
@@ -842,77 +842,91 @@ void Lowering::LowerSIMD(GenTreeSIMD* simdNode)
...
-                    }
-                }
+    switch (newIntrinsicId)
+    {
+        case NI_SSE_COMISS:
+        case NI_SSE_UCOMISS:
+        case NI_SSE2_COMISD:
+        case NI_SSE2_UCOMISD:
+            // Using the preferred condition saves a branch so it's probably better
+            // to always use it, even if that means loosing containment.
```
Comment (CarolEidt) [inline src/jit/lowerxarch.cpp]: I am missing where this handles the case where it needs to lose containment

File: src/jit/lower.cpp [hunk 8]
```diff
@@ -2932,6 +2932,107 @@ GenTree* Lowering::LowerJTrue(GenTreeOp* jtrue)
...
+    GenTreeCC* cc = nullptr;
+
+    // Next may be null if `node` is not used. In that case we don't need to generate a SETCC node.
+    if (next != nullptr)
+    {
+        if (next->OperIs(GT_JTRUE))
+        {
+            // This should always be true, otherwise it means that JTRUE's relop is somewhere
+            // before `node` and that shouldn't happen. Still, if it happens it's not our problem,
+            // it simply means that `node` isn't used and we don't have to do anything.
```
Comment (CarolEidt) [inline src/jit/lower.cpp]: This comment is a bit confusing, especially wrt to what "This" is referring to (I mistakenly read it as refering to the condition above at first). I would reword as:
```
\\ If the instruction immediately following 'relop', i.e. 'next' is a conditional branch, it should always have 'relop' as its 'op1'. 
\\ If it doesn't, then we have improperly constructed IL (the setting of a condition code should always immediately
\\ precede its use, since the JIT doesn't track dataflow for condition codes). Still, if it happens ...
```
Or something to that effect. It might even be worth mentioning that restriction (i.e. that the setting and use of condition
codes must always be contiguous) in the header.

File: src/jit/lowerxarch.cpp (hunk 7 above)
Comment (mikedn) [reply src/jit/lowerxarch.cpp]: If we have `[mem], reg` we may be able to contain the mem operand by swapping the operands to have `reg, [mem]`. But that also requires changing the comparison condition and unfortunately some FP conditions have complicated codegen on xarch. So the code below gives priority to "preferrable" FP conditions by blocking operands swapping and thus indirectly blocking containment in some cases.

I'll try to expand this comment a bit tomorrow, together with the other comment improvement suggestion you've made.

Comment (CarolEidt) [reply src/jit/lowerxarch.cpp]: I think I now see my point of confusion. When we get here, we haven't yet done containment analysis, correct? That's why we don't need to "undo" anything.

Comment (mikedn) [reply src/jit/lowerxarch.cpp]: Yes. Hopefully the new comments is more clear.

**Comment by CarolEidt** [comment]
The x86 failures are the same as those from #26551, and are, in addition, in tests that have no HW intrinsics.

**Comment by GrabYourPitchforks** [comment]
This is awesome to see! Thanks @mikedn for driving this - much appreciated. :)
